### ISSUE TITLE
Compile with MKL and intel's compiler

### ISSUE THREAD

**Issue by SanPen** [coin-or/Cbc#252]
Hi,

I am trying to compile Cbc with MKL and intel's compiler under Ubuntu 18.04 (x64_86)

The `./configure` routine is:

> ./configure -C 
> --enable-cbc-parallel 
> --enable-static 
> --disable-shared  
> --without-gcc 
> --prefix=/mnt/sdb1/CPP_libs/Cbc/build 
> --includedir=/opt/intel/compilers_and_libraries/linux/mkl/include 
> --libdir=/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64  
> --with-blas-incdir 
> --with-blas-lib 
> --with-lapack-incdir 
> --with-lapack-lib 
> FC=ifort 
> CXX=icc 
> CC=icc

The configuration itself goes fine, but when compiling with a simple `make`, at some point it fails with the message:

> *** No rule to build the objective 'yes', necessary for 'clp'.  Stop.

I wonder if the configuration is appropriate, and if you could provide a suitable one.

**Comment by svigerske** [comment]
Hmm, no idea how you get there.

The configure flags instruct the buildsystem to install the Cbc headers in the MKL include dir and the Cbc libraries into the MKL library directory. The `--with-blas-*` and `--with-lapack-*` are just ignored in best case, as they come without value. gfortran may still be used for Fortran 77 code.

You probably wanted something like
```
./configure -C
--enable-cbc-parallel
--enable-static
--disable-shared
--prefix=/mnt/sdb1/CPP_libs/Cbc/build
--with-blas="-L/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64 -lmkl_core -lmkl_intel_lp64 -lmkl_gnu_thread -fopenmp -lm"
F77=ifort
CXX=icc
CC=icc
```
Further, there can be a problem if the linker picks up the static MKL libraries, because they have circular references and something like `-Wl,--start-group / -Wl,--end-group` does not work with libtool.

**Comment by SanPen** [comment]
Thanks for your answer.

This worked, but it generates the dynamically linked version of Cbc. 
However since I am compiling Cbc to be shipped with PuLP (a python MIP library), I need the statically linked version.

> ./configure -C 
> --enable-cbc-parallel 
> --enable-static 
> --disable-shared 
> --without-gcc 
> --prefix=/mnt/sdb1/CPP_libs/Cbc-2.10/build 
> --with-lapack="-L/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64 -static -mkl=parallel -Wl,--start-group -lmkl_intel_lp64 -lmkl_core -lmkl_intel_thread -lpthread -lm -lmkl_gnu_thread -Wl,--end-group -fopenmp " 
> F77=ifort 
> FC=ifort 
> CXX=icc
> CC=icc

the same one but using `--with-blas` instead of `--with-lapack` does the same. 

Any insight on how to force a statically linked version with MKL?

**Comment by svigerske** [comment]
I'm not so optimistic that this would be possible. See also https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=159760&repeatmerged=yes

They mention one workaround, though, maybe that will actually work (I haven't tried). The thing do to would be to use something like
```
--with-blas="-L/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64 -Wl,--start-group,libmkl_core.a,libmkl_intel_lp64.a,libmkl_gnu_thread.a,--end-group -fopenmp -lm"
```

**Comment by tkralphs** [comment]
I have personally tried to generate fully static Cbc binaries on Linux before for the same purpose and after some investigation, I gave up. There are lot of recipes out there, but I always ended up with some dependence on some dynamic library no matter what I tried. Let us know if you succeed, I would like to deploy such statically linked binaries for general use. The AMPL guys  seem to have managed to do it with some CMake incantations, but I couldn't replicate it.

**Comment by tkralphs** [comment]
FYI, there was some discussion here: https://github.com/coin-or/pulp/issues/196

**Comment by SanPen** [comment]
I have sucedded to compile the static Cbc executable both with `gcc` and `icc`.
You can find them [here](https://github.com/SanPen/GridCal/tree/master/src/GridCal/ThirdParty/pulp/solverdir/cbc/linux/64)

The [`cbc` executable of 18 MB](https://github.com/SanPen/GridCal/blob/master/src/GridCal/ThirdParty/pulp/solverdir/cbc/linux/64/cbc) was compiled with `icc`. No MKL, just compiling BLAS and LAPACK with icc, following the normal compilation procedure.
[The one of 6 MB](https://github.com/SanPen/GridCal/blob/master/src/GridCal/ThirdParty/pulp/solverdir/cbc/linux/64/cbc.bak) was compiled with `gcc`. Compiling BLAS and LAPACK with gcc, following the normal compilation procedure.

Both with the multi-threading option.

However! I find the icc one to be just 3% faster (wich may be a negligible performance difference in some cases). It would be fantastic that when linking MKL or OpenBlas, the Cbc code would use multi-threaded linear algebra (matrix products and posibly Cholesky factorization) as much as possible. Maybe an implementation of [this iseas](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.6.2640&rep=rep1&type=pdf) would be desired.

**Comment by tkralphs** [comment]
Could you share details? What was the exact configure command? I take it you did not include any third party projects?

**Comment by SanPen** [comment]
Hi, 

The steps are:

> git clone --branch=stable/2.10 https://github.com/coin-or/Cbc Cbc-2.10

> cd Cbc-2.10

> git clone --branch=stable/0.8 https://github.com/coin-or-tools/BuildTools/

> bash BuildTools/get.dependencies.sh fetch

> ./configure -C --enable-cbc-parallel --enable-static --disable-shared --prefix=/mnt/sdb1/CPP_libs/Cbc-2.10/build

> make

> make install

Let me know if you succeed

**Comment by tkralphs** [comment]
What Linux distro are you using? With coinbrew on Ubunutu, if I do
```
./coinbrew build Cbc --enable-static --disable-shared --no-third-party --build-dir build-cbc-static ADD_CXXFLAGS="-static -static-libgcc -static-libstdc++" --disable-zlib
```
then I still have
```
~/Projects/coinbrew/dist/bin > ldd ./cbc
        linux-vdso.so.1 (0x00007ffff80ad000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f83f9e70000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f83f9a70000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f83fac00000)
```
It seems that forcing `libm` to link statically is the key and not that straightforward, at least in the context of our build system. [This Stackoverflow post](https://stackoverflow.com/questions/26277283/gcc-linking-libc-static-and-some-other-library-dynamically-revisited) addresses this exact question. I do have `/usr/lib/x86_64-linux-gnu/libm.a` so I could hack the build system to link directly to it, but that is not a portable solution and it's not clear that it would eliminate dynamic linking of the other libraries. I tried modifying `coin.m4` to do `-Wl,-Bstatic -lm`, but that didn't seem to have any effect. 

So I'm still running into this wall and would like to know what you did to make this work. I guess it must be something about your particular setup (like maybe you don't have `libm.so` or something).

**Comment by SanPen** [comment]
Hi, I have Ubuntu 18.04.
I did not use coinbrew to compile, just the list of commands that I posted.
I will re-check, but I managed to make it work like that a couple of times.

**Comment by tkralphs** [comment]
`coinbrew` just passes the arguments through to the respective `configure` commands of individual projects, so all but `--no-third-party --build-dir build-cbc-static` will get passed through. It's strange that you succeeded so easily, since the Stackoverflow post linked above seems to indicate that the linking of `libm` is an issue. Anyway, let me know what you find.

**Comment by SanPen** [comment]
Well `./configure`, `make`, `make install` is probaly the first thing I learnt at the linux world way back when. I encourage you to compile using that instead of other methods. I did succeed whith this in two different computers with Ubuntu 18.04. I'll try in a fresh VM.

**Comment by tkralphs** [comment]
I wrote the `coinbrew` script and am also one of the main developers of the [BuildTools](https://github.com/coin-or-tools/BuildTools), so I'm very familiar with what `coinbrew` is doing under the hood. It is doing `configure`, `make`, `make install` for each of the dependencies. The purpose of `coinbrew` is to manage all of the dependencies and the build process once the root `configure` script goes away, which will happen "soon." That is why we're encouraging users to move to `coinbrew` as soon as possible to avoid breaking your build process down the road.  

If you succeed in a fresh VM, maybe you can give details of that so I can try to start trying to figure out where the issue might be. I am also on Ubuntu 18.04, but could have something to do with some packages I've installed, who knows. I can give a try building on Travis with a fresh VM there and see if that works.

**Comment by SanPen** [comment]
Then, I am really confused how is that you cannot get to compile the static version!

I suggested the ./configure way because it'll tell you about all the missing dependencies.

**Comment by tkralphs** [comment]
Actually, could you just post `config.log` and `config.status` from both the root directory and the `Cbc` subdirectory? I may also eventually want the full log of the build, but let's start with those files.

**Comment by SanPen** [comment]
Just compiled
[config.log](https://github.com/coin-or/Cbc/files/3566834/config.log)
[config.status.txt](https://github.com/coin-or/Cbc/files/3566835/config.status.txt)

**Comment by tkralphs** [comment]
Can you attach the ones from the `Cbc/` subdirectory and also `CoinUtils/`?

**Comment by SanPen** [comment]
From the CbC folder:
[config.log](https://github.com/coin-or/Cbc/files/3570372/config.log)
[config.status.txt](https://github.com/coin-or/Cbc/files/3570374/config.status.txt)

**Comment by SanPen** [comment]
By the way the original thread issue was how to link MKL, so it would be great if the developers could provide insight here.

**Comment by svigerske** [comment]
I tried what [I mentioned above](https://github.com/coin-or/Cbc/issues/252#issuecomment-525921601), and this helps to work with circular dependencies within the MKL libraries, but the `-Wl,--...` flags are unfortunately not stored by libtool in the dependencies of the .la files, so that the Lapack-dependency of, e.g., CoinUtils, is not carried on until Clp and Cbc and linking then fails there. I haven't tried whether putting this directly into LDFLAGS would actually work. Another workaround would be to just look at the final linkline that is produced by libtool for, e.g, the cbc executable, and modify this accordingly, e.g., add a `-static`.

**Comment by tkralphs** [comment]
OK, yeah, I guess this is a side conversation that should in another issue. But just to close it out, I tried precisely the same recipe as you (git clone, root configure, building cbc in parallel, everything precisely the same) and got
```
~/tmp/Cbc-2.10/bin > ldd ./cbc
        linux-vdso.so.1 (0x00007fffd2944000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ff7a3250000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007ff7a3040000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007ff7a2e20000)
        libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007ff7a2b90000)
        libgfortran.so.4 => /usr/lib/x86_64-linux-gnu/libgfortran.so.4 (0x00007ff7a27b0000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ff7a25a0000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ff7a2200000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ff7a1e60000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ff7a1c40000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ff7a1840000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff7a4000000)
        libquadmath.so.0 => /usr/lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007ff7a1600000)
```
Maybe you can send me the full output in the other issue that will now open.

**Comment by SanPen** [comment]
Just to add to the conversation; I managed to compile with MKL in windows after a lot of VS project property clicking. So it should be possible in Linux as well but I lack the knowledge to configure all the projects as they should :(

**Comment by pzdump** [comment]
I'm also interested in building CBC with MKL but using gcc since i don't have a icc.
How can you do that with coinbrew script?

every try i did it d either failed with
configure: error: Could not find dsyev in Lapack
or coinbrew just picking up standard blas/lapack.

compile command with not finding dsyev was:
../coinbrew build install Cbc --enable-cbc-parallel --verbosity=4 --test --prefix=/opt/msp/cbc-trunk-2020-01-24_mkl --with-lapack="-L/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64 -mkl=parallel -Wl,--start-group -lmkl_intel_lp64 -lmkl_core -lmkl_intel_thread -lpthread -lm -lmkl_gnu_thread -Wl,--end-group -fopenmp " --with-blas="-L/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64 -mkl=parallel -Wl,--start-group -lmkl_intel_lp64 -lmkl_core -lmkl_intel_thread -lpthread -lm -lmkl_gnu_thread -Wl,--end-group -fopenmp " --no-prompt >buildlog.txt 2>&1

Any hint how to do that with coinbrew?
Thanks

**Comment by svigerske** [comment]
Building against static MKL libs is difficult, as the `-Wl,--start-group` and `-Wl,--end-group` are moved around by libtool. If you have shared MKL libs somewhere where gcc is looking anyway, then sequential MKL should be picked up automatically by configure. Otherwise something like `--with-lapack="-L/opt/intel/compilers_and_libraries/linux/mkl/lib/intel64 -lmkl_intel_lp64 -lmkl_core -lmkl_intel_thread -lpthread  -fopenmp"` should work. If not, check the corresponding config.log.
